Archiver/src/EngineEnvironment.cpp

```diff
@@ -0,0 +1,192 @@
+    m_pDeviceReflection = std::make_unique<DeviceObjectReflection>(m_pSerializationDevice, m_pShaderStreamFactory, m_Desc.DeviceBits);
+}
+
+EngineEnvironment::~EngineEnvironment()
+{}
+
+void EngineEnvironment::Initialize(int argc, char** argv)
+{
+    argparse::ArgumentParser Parser{"JSON Archiver"};
+
+    Parser.add_argument("-o", "--output")
+        .required()
+        .help("Output Binary Archive");
+
+    Parser.add_argument("-s", "--shader_dir")
+        .default_value(std::string("."))
+        .help("Shaders directory");
+
+    Parser.add_argument("-c", "--config")
+        .default_value("")
+        .help("Path to config");
+
+    Parser.add_argument("-dx11")
+        .default_value(false)
+        .implicit_value(true);
+
+    Parser.add_argument("-dx12")
+        .default_value(false)
+        .implicit_value(true);
+
+    Parser.add_argument("-vulkan")
+        .default_value(false)
+        .implicit_value(true);
+
+    Parser.add_argument("-opengl")
+        .default_value(false)
+        .implicit_value(true);
+
+    Parser.add_argument("-metal")
+        .default_value(false)
+        .implicit_value(true);
+
+    Parser.add_argument("-i", "--input")
+        .required()
+        .remaining()
+        .help("Input Json Archive");
+
+    try
+    {
+        Parser.parse_args(argc, argv);
+    }
+    catch (const std::runtime_error& err)
+    {
+        std::cerr << err.what() << std::endl;
+        std::cerr << Parser;
+        std::exit(1);
+    }
+
+    auto GetDeviceBitsFromParser = [](argparse::ArgumentParser const& Parser) {
+        RENDER_DEVICE_TYPE_FLAGS DeviceBits = {};
+        if (Parser.get<bool>("-dx11"))
+            DeviceBits |= RENDER_DEVICE_TYPE_FLAG_D3D11;
+        if (Parser.get<bool>("-dx12"))
+            DeviceBits |= RENDER_DEVICE_TYPE_FLAG_D3D12;
+        if (Parser.get<bool>("-vulkan"))
+            DeviceBits |= RENDER_DEVICE_TYPE_FLAG_VULKAN;
+        if (Parser.get<bool>("-metal"))
+            DeviceBits |= RENDER_DEVICE_TYPE_FLAG_METAL;
+        if (Parser.get<bool>("-opengl"))
+            DeviceBits |= RENDER_DEVICE_TYPE_FLAG_GL;
+        return DeviceBits;
+    };
+
+    auto DeviceBits     = GetDeviceBitsFromParser(Parser);
+    auto ShaderFilePath = Parser.get<std::string>("--shader_dir");
+    auto ConfigFilePath = Parser.get<std::string>("--config");
+    auto OutputFilePath = Parser.get<std::string>("--output");
+    auto InputFilePaths = Parser.get<std::vector<std::string>>("--input");
+
+
+    if (!m_pEnvironment)
+    {
+        EngineEnvironmentDesc Desc = {};
+
+        Desc.DeviceBits      = DeviceBits;
+        Desc.ShadersFilePath = std::move(ShaderFilePath);
+        Desc.ConfigFilePath  = std::move(ConfigFilePath);
+        Desc.OuputFilePath   = std::move(OutputFilePath);
+        Desc.InputFilePaths  = std::move(InputFilePaths);
+
+        m_pEnvironment = new EngineEnvironment{Desc};
```

Comment:
NEVER use new/delete with raw pointers.
This must be at least unique_ptr, use make_unique.

This singleton should be removed completely I am sure.